File: Basic/Chapter5/IntervalDP.py
"""
Interval DP
"""
"""
282. 石子合并
https://www.acwing.com/problem/content/284/
"""

n = int(input())
nums = list(map(int, input().split()))
f = [[1e9] * n for _ in range(n)]
for i in range(n):
    f[i][i] = 0
s = [0] * (n + 1)
for i in range(1, n + 1):
    s[i] = s[i - 1] + nums[i - 1]

for interval in range(2, n + 1):
    # left: i, right: i + interval - 1
    # enumerate left start
    for i in range(n - interval + 1):
        l, r = i, i + interval - 1
        for k in range(l, r):
            f[l][r] = min(f[l][r], f[l][k] + f[k + 1][r] + s[r + 1] - s[l])
print(f[0][n - 1])

File: Basic/Chapter5/IntervalDP.cpp
#include <iostream>

using namespace std;

const int N = 310;
int n, s[N], f[N][N];

int main() {
  cin >> n;
  for (int i = 1; i <= n; ++i) {
    int x;
    cin >> x;
    s[i] = s[i - 1] + x;
  }

  for (int interval = 2; interval <= n; ++interval) {
    for (int i = 0; i + interval <= n; ++i) {
      int l = i, r = i + interval - 1;
      f[l][r] = 1e9;
      for (int k = l; k < r; ++k) {
        f[l][r] = min(f[l][r], f[l][k] + f[k + 1][r] + s[r + 1] - s[l]);
      }
    }
  }

  cout << f[0][n - 1];
  return 0;
}